matrix_factorization/training.cu

```diff
@@ -31,47 +30,90 @@ float *initialize_normal_array(int size) {
+    CudaDenseMatrix* P_device_target = new CudaDenseMatrix(user_count, n_factors, P);
+    CudaDenseMatrix* Q_device_target = new CudaDenseMatrix(item_count, n_factors, Q);
 
     // Create the errors
     float *errors_device;
     cudaMalloc(&errors_device, matrix->nonzeros * sizeof(float));
-    // TODO: see if we can remove this
-    cudaMemset(errors_device, 0, matrix->nonzeros * sizeof(float));
 
     // Create the total losses
     float *losses_device;
     cudaMalloc(&losses_device, n_iterations * sizeof(float));
 
+    // Create the bias arrays
+    float *user_bias = initialize_normal_array(user_count);
+    float *item_bias = initialize_normal_array(item_count);
+    *user_bias_ptr = user_bias;
+    *item_bias_ptr = item_bias;
+    
+    float *user_bias_device;
+    cudaMalloc(&user_bias_device, user_count * sizeof(float));
+    cudaMemcpy(user_bias_device, user_bias, user_count * sizeof(float), cudaMemcpyHostToDevice);
+
+    float *item_bias_device;
+    cudaMalloc(&item_bias_device, item_count * sizeof(float));
+    cudaMemcpy(item_bias_device, item_bias, item_count * sizeof(float), cudaMemcpyHostToDevice);
+
+    // Create bias targets
+    float *user_bias_target, *item_bias_target;
+    cudaMalloc(&user_bias_target, user_count * sizeof(float));
+    cudaMemcpy(user_bias_target, user_bias, user_count * sizeof(float), cudaMemcpyHostToDevice);
+    cudaMalloc(&item_bias_target, item_count * sizeof(float));
+    cudaMemcpy(item_bias_target, item_bias, item_count * sizeof(float), cudaMemcpyHostToDevice);
+
     // Dimensions
     int n_threads = 32;
     dim3 dim_block(n_threads);
-    dim3 dim_grid_sgd(matrix->rows / n_threads + 1);
+    dim3 dim_grid_sgd(user_count / n_threads + 1);
     dim3 dim_grid_loss(matrix->nonzeros / n_threads + 1);
 
     // Training loop
+    cudaError_t lastError;
     for (int i = 0; i < n_iterations; ++i) {
-        calculate_loss_gpu(P_device, Q_device, n_factors, matrix->rows, matrix->cols, matrix->nonzeros,
-                           matrix, errors_device);
+        // Calculate initial error per each rating
+        calculate_loss_gpu(P_device, Q_device, n_factors, user_count, item_count, matrix->nonzeros, matrix,
+                           errors_device, user_bias_device, item_bias_device);
+
+        // Run single iteration of SGD
         sgd_update<<<dim_grid_sgd, dim_block>>>(matrix->indptr, matrix->indices, P_device->data, Q_device->data,
                                                 P_device_target->data, Q_device_target->data, n_factors, errors_device,
-                                                matrix->rows, matrix->cols, learning_rate);
+                                                user_count, item_count, learning_rate, user_bias_device, item_bias_device,
+                                                user_bias_target, item_bias_target, learning_rate, learning_rate);
+        lastError = cudaGetLastError();
+        if(cudaSuccess != lastError) {
```

Comment:
Can we check for error after the training loop? Feel like it might slow things down. Or maybe we can do it in the first iteration (to see if everything seems to be working) and one at the end?

Reply:
if you check after loop, you will only get an error from last kernel call, ie total_loss